src/main/java/com/gooddata/http/client/GoodDataHttpClient.java

```diff
@@ -94,35 +92,72 @@
 
     private final SSTRetrievalStrategy sstStrategy;
 
-    private final HttpContext context;
+    /** Host performing authentication - eg. issuing TT tokens */
+    private final HttpHost authHost;
+
+    /** this lock is used to ensure that no threads will try to send requests while authentication is performed */
+    private final ReadWriteLock rwLock = new ReentrantReadWriteLock();
+
+    /** this lock guards that only one thread enters the authentication (obtaining TT/SST) section */
+    private final Lock authLock = new ReentrantLock();
+
+    /** current SST (or null if not yet obtained) */
+    private String sst;
+
+    /** TT to be set into the header (or null if not yet obtained) */
+    private String tt;
+
 
     /**
      * Construct object.
      * @param httpClient Http client
      * @param sstStrategy super-secure token (SST) obtaining strategy
      */
+    @Deprecated
```

Comment:
It would be fine to describe in javadoc, what should user use instead of these deprecated ones.